insert_by_period/macros/get_period_sql.sql

```diff
@@ -19,3 +19,22 @@
   ) target_cols
 
 {%- endmacro %}
+
+
+{% macro bigquery__get_period_sql(target_cols_csv, sql, timestamp_field, period, start_timestamp, stop_timestamp, offset) -%}
+
+  {%- set period_filter -%}
+    ({{timestamp_field}} >  cast(cast(timestamp('{{start_timestamp}}') as datetime) + interval {{offset}} {{period}} as timestamp) and
+     {{timestamp_field}} <= cast(cast(timestamp('{{start_timestamp}}') as datetime) + interval {{offset}} {{period}} + interval 1 {{period}} as timestamp) and
+     {{timestamp_field}} <  cast('{{stop_timestamp}}' as timestamp))
+  {%- endset -%}
```

Comment:
Have you thought about the implications of `datetime` and `timestamp` in BQ vs. `timestamp` in Snowflake?

My understanding is that `timestamp` is aware in BQ and `datetime` is naive (whereas `timestamp` is naive by default in Snowflake, but can also be configured to have different behavior).

We've been a bit inconsistent in how we've treated date/time related fields in various places in dbt which makes this a bit harder to reason about what the best behavior should be in this instance.

For example, here's how dbt-snowflake and dbt-bigquery compare in a couple different ways:

<img width="772" alt="image" src="https://github.com/dbt-labs/dbt-labs-experimental-features/assets/44704949/4d0692f2-a036-4feb-9f72-2ebe44c2bf6d">

<img width="774" alt="image" src="https://github.com/dbt-labs/dbt-labs-experimental-features/assets/44704949/5819808e-dc16-4739-b9cf-72a57dfff9b4">

Reply:
What are you suggesting then here?

Should we update `start_date` and `stop_date` to be `start_timestamp` and `stop_timestamp` then? And tell people that those timestamps need to have the same Offset/TZ-awareness as the field `timestamp_field`?

And if we break the current column names, then let's rename `timestamp_field` to `timestamp_expression` as well to be more explicit.

Reply:
All of those would make sense to me!

The primary thing I am trying to call out is that most people overlook how important it is that everything has the same same TZ-awareness / UTC offset. Otherwise, it's very possible to get different output than expected (and not actually notice it).

We've also been accidentally inconsistent our various time-related macros _within_ dbt (as seen comparing dbt-bigquery with dbt-snowflake in the images above), which doesn't help matters.

### Breaking changes

Our `README` says:
> In all cases, these are demo projects, not intended as ready-to-use packages

So that seems to give us the license to merge this PR as-is. And also the license to introduce breaking changes.

But we can be almost certain that there would be folks that didn't notice that caveat and are just installing this code from the `main` branch. And such changes might break their workflows.

### Backwards-compatibility

There's prior art for changing the official parameter names while retaining backwards-compatibility with the old parameter names. Going that route might be a better option if we do the following:

- rename `start_date` -> `start_timestamp`
- rename `stop_date` -> `stop_timestamp`
- rename `timestamp_field` to `timestamp_expression`

It seems like renames ~could~ should happen in a separate follow-up PR to make them easier to review.

As long as we don't make any backwards-incompatible changes, I'd be okay with shipping this PR as-is.

Reply:
Thanks Doug. I just pushed the change removing the commented code.

What's next? Are you OK to approve it so I merge the code?
Or can I merge it without approval?

As you said we can raise another issue/PR for the renaming.

Reply:
I'm going to approve shortly 👍